src/main.rs

```diff
@@ -263,20 +242,24 @@ impl VideoPlayer {
     }
 
     pub fn start(&mut self) {
-        let callback = || glib::idle_add(ui_action_handle);
         let sender = self.sender.clone();
-        let receiver = self.player_receiver.clone();
-
-        thread::spawn(move || {
-            while let Ok(event) = receiver.recv() {
-                // if let PlayerEvent::EndOfPlaylist = event {
-                //     sender.send(UIAction::Quit).unwrap();
-                //     callback();
-                //     break;
-                // }
-                sender.send(UIAction::ForwardedPlayerEvent(event)).unwrap();
-                callback();
-            }
+        let receiver = self.player_receiver.take().expect("No player channel receiver");
+        receiver.attach(None, move |event| {
+            sender.send(UIAction::ForwardedPlayerEvent(event)).unwrap();
+
+            let mut cont = true;
+            with_video_player!(player {
+                while let Ok(action) = player.receiver.try_recv() {
```

Comment:
Why do you have another receiver here instead of passing data into the glib sender directly? How do you ensure that the glib receiver wakes up every time these other receivers have data?

Reply:
Can't remember right now why I did it like this with the crossbeam channels... I need to revise this :)

Reply:
We need a two-way communication between the UI and the player, can that be done with a single channel? The player should be able to notify the UI without expecting a response too (iirc).

Reply:
But this is two times from somewhere into the main loop (the callback itself is a receive, and then you call recv from there again). For sending things back you need another channel, yes.

Reply:
They're different receivers. I don't see what's wrong here :)

Reply:
Polling a receiver from inside another receiver seems like an anti-pattern and I don't see how you ensure that the glib channel always wakes up when that other receiver has items available.

Reply:
Oh my, what monster did I create. OK there's some cleanup to do there indeed.